File: game.cpp
#include <SFML/Graphics.hpp>
#include <SFML/Window.hpp>
#include <iostream>
#include <vector>

using namespace std;

// Game constants
const int windowWidth = 800;
const int windowHeight = 600;
const int paddleWidth = 120;
const int paddleHeight = 15;
const int ballRadius = 8;
const int brickWidth = 75;
const int brickHeight = 25;
const int brickPadding = 5;
const int brickOffsetTop = 60;
const int brickOffsetLeft = 35;
const int brickRows = 6;
const int brickCols = 10;

// Game objects
int score = 0;
int lives = 3;

sf::RectangleShape paddle(sf::Vector2f(paddleWidth, paddleHeight));
sf::CircleShape ball(ballRadius);
std::vector<std::vector<sf::RectangleShape>> bricks(brickRows, std::vector<sf::RectangleShape>(brickCols));

// Input handling
bool rightPressed = false, leftPressed = false;
sf::Vector2i mousePosition(0, 0);

// Ball and Paddle movements
float ballDx = 5.f, ballDy = -5.f;
float paddleSpeed = 8.f;

// Function to reset ball position
void resetBall() {
    ball.setPosition(windowWidth / 2.f, windowHeight - 50.f);
    ballDx = 5.f * (rand() % 2 == 0 ? 1 : -1);
    ballDy = -5.f;
}

// Function to initialize bricks
void initBricks() {
    for (int r = 0; r < brickRows; ++r) {
        for (int c = 0; c < brickCols; ++c) {
            bricks[r][c].setSize(sf::Vector2f(brickWidth, brickHeight));
            bricks[r][c].setFillColor(sf::Color(rand() % 256, rand() % 256, rand() % 256));
            bricks[r][c].setPosition(c * (brickWidth + brickPadding) + brickOffsetLeft, r * (brickHeight + brickPadding) + brickOffsetTop);
        }
    }
}

// Function to handle the drawing of the game
void drawGame(sf::RenderWindow &window) {
    // Clear the window
    window.clear(sf::Color::Black);

    // Draw bricks
    for (int r = 0; r < brickRows; ++r) {
        for (int c = 0; c < brickCols; ++c) {
            if (bricks[r][c].getSize().x != 0) {
                window.draw(bricks[r][c]);
            }
        }
    }

    // Draw paddle
    window.draw(paddle);

    // Draw ball
    window.draw(ball);

    // Display the current score and lives
    sf::Font font;
    if (!font.loadFromFile("arial.ttf")) {
        std::cerr << "Error loading font" << std::endl;
    }

    sf::Text scoreText("Score: " + std::to_string(score), font, 20);
    scoreText.setFillColor(sf::Color::White);
    scoreText.setPosition(10.f, 10.f);
    window.draw(scoreText);

    sf::Text livesText("Lives: " + std::to_string(lives), font, 20);
    livesText.setFillColor(sf::Color::White);
    livesText.setPosition(windowWidth - 100.f, 10.f);
    window.draw(livesText);

    window.display();
}

// Function to handle collisions with bricks
void collisionDetection() {
    for (int r = 0; r < brickRows; ++r) {
        for (int c = 0; c < brickCols; ++c) {
            sf::RectangleShape &brick = bricks[r][c];
            if (brick.getSize().x != 0) {
                if (ball.getGlobalBounds().intersects(brick.getGlobalBounds())) {
                    ballDy = -ballDy;
                    brick.setSize(sf::Vector2f(0, 0)); // Destroy brick
                    score += 10;

                    // Check if all bricks are destroyed
                    if (score == brickRows * brickCols * 10) {
                        std::cout << "You Win!" << std::endl;
                        lives = 0; // End the game
                    }
                }
            }
        }
    }
}

// Function to handle paddle movement
void movePaddle() {
    if (rightPressed && paddle.getPosition().x + paddleWidth < windowWidth) {
        paddle.move(paddleSpeed, 0);
    }
    if (leftPressed && paddle.getPosition().x > 0) {
        paddle.move(-paddleSpeed, 0);
    }
    // Mouse control
    paddle.setPosition(mousePosition.x - paddleWidth / 2, paddle.getPosition().y);
}

// Function to update the ball's position
void moveBall() {
    ball.move(ballDx, ballDy);

    // Ball collision with the left and right walls
    if (ball.getPosition().x <= 0 || ball.getPosition().x + ballRadius * 2 >= windowWidth) {
        ballDx = -ballDx;
    }

    // Ball collision with the top wall
    if (ball.getPosition().y <= 0) {
        ballDy = -ballDy;
    }

    // Ball collision with the paddle
    if (ball.getGlobalBounds().intersects(paddle.getGlobalBounds())) {
        ballDy = -ballDy;
    }

    // Ball falls below the paddle
    if (ball.getPosition().y + ballRadius * 2 > windowHeight) {
        lives--;
        if (lives == 0) {
            std::cout << "Game Over! Final Score: " << score << std::endl;
            exit(0);
        } else {
            resetBall();
        }
    }
}

int main() {
    sf::RenderWindow window(sf::VideoMode(windowWidth, windowHeight), "Brick Breaker Game");

    // Initialize game objects
    resetBall();
    initBricks();

    // Main game loop
    while (window.isOpen()) {
        sf::Event event;
        while (window.pollEvent(event)) {
            if (event.type == sf::Event::Closed) {
                window.close();
            }

            if (event.type == sf::Event::KeyPressed) {
                if (event.key.code == sf::Keyboard::Right) {
                    rightPressed = true;
                }
                if (event.key.code == sf::Keyboard::Left) {
                    leftPressed = true;
                }
            }

            if (event.type == sf::Event::KeyReleased) {
                if (event.key.code == sf::Keyboard::Right) {
                    rightPressed = false;
                }
                if (event.key.code == sf::Keyboard::Left) {
                    leftPressed = false;
                }
            }

            if (event.type == sf::Event::MouseMoved) {
                mousePosition = sf::Mouse::getPosition(window);
            }
        }

        // Game logic
        moveBall();
        movePaddle();
        collisionDetection();

        // Draw everything
        drawGame(window);
    }

    return 0;
}

File: game.js
const canvas = document.getElementById('gameCanvas');
const ctx = canvas.getContext('2d');

// Game variables
let score = 0;
let lives = 3;
let gameRunning = false;
let gameStarted = false;

// Paddle
const paddle = {
    x: canvas.width / 2 - 60,
    y: canvas.height - 30,
    width: 120,
    height: 15,
    speed: 8
};

// Ball
const ball = {
    x: canvas.width / 2,
    y: canvas.height - 50,
    radius: 8,
    dx: 5,
    dy: -5,
    speed: 5
};

// Bricks
const brickRows = 6;
const brickCols = 10;
const brickWidth = 75;
const brickHeight = 25;
const brickPadding = 5;
const brickOffsetTop = 60;
const brickOffsetLeft = 35;

let bricks = [];
const brickColors = ['#ff6b6b', '#4ecdc4', '#45b7d1', '#96ceb4', '#feca57', '#ff9ff3'];

// Initialize bricks
function initBricks() {
    bricks = [];
    for (let r = 0; r < brickRows; r++) {
        bricks[r] = [];
        for (let c = 0; c < brickCols; c++) {
            bricks[r][c] = {
                x: c * (brickWidth + brickPadding) + brickOffsetLeft,
                y: r * (brickHeight + brickPadding) + brickOffsetTop,
                status: 1,
                color: brickColors[r]
            };
        }
    }
}

// Input handling
let rightPressed = false;
let leftPressed = false;
let mouseX = 0;

document.addEventListener('keydown', keyDownHandler);
document.addEventListener('keyup', keyUpHandler);
canvas.addEventListener('mousemove', mouseMoveHandler);

function keyDownHandler(e) {
    if (e.key === 'Right' || e.key === 'ArrowRight') {
        rightPressed = true;
    } else if (e.key === 'Left' || e.key === 'ArrowLeft') {
        leftPressed = true;
    } else if (e.key === ' ' || e.key === 'Spacebar') {
        e.preventDefault();
        if (!gameStarted) {
            startGame();
        }
    }
}

function keyUpHandler(e) {
    if (e.key === 'Right' || e.key === 'ArrowRight') {
        rightPressed = false;
    } else if (e.key === 'Left' || e.key === 'ArrowLeft') {
        leftPressed = false;
    }
}

function mouseMoveHandler(e) {
    const rect = canvas.getBoundingClientRect();
    mouseX = e.clientX - rect.left;
}

// Drawing functions
function drawBall() {
    ctx.beginPath();
    ctx.arc(ball.x, ball.y, ball.radius, 0, Math.PI * 2);
    ctx.fillStyle = '#ffffff';
    ctx.fill();
    ctx.shadowColor = '#ffffff';
    ctx.shadowBlur = 10;
    ctx.closePath();
    ctx.shadowBlur = 0;
}

function drawPaddle() {
    ctx.beginPath();
    ctx.roundRect(paddle.x, paddle.y, paddle.width, paddle.height, 8);
    ctx.fillStyle = '#ffffff';
    ctx.fill();
    ctx.closePath();
}

function drawBricks() {
    for (let r = 0; r < brickRows; r++) {
        for (let c = 0; c < brickCols; c++) {
            if (bricks[r][c].status === 1) {
                const brick = bricks[r][c];
                ctx.beginPath();
                ctx.roundRect(brick.x, brick.y, brickWidth, brickHeight, 5);
                ctx.fillStyle = brick.color;
                ctx.fill();
                ctx.strokeStyle = '#ffffff';
                ctx.lineWidth = 1;
                ctx.stroke();
                ctx.closePath();
            }
        }
    }
}

// Collision detection
function collisionDetection() {
    for (let r = 0; r < brickRows; r++) {
        for (let c = 0; c < brickCols; c++) {
            const brick = bricks[r][c];
            if (brick.status === 1) {
                if (ball.x > brick.x && ball.x < brick.x + brickWidth &&
                    ball.y > brick.y && ball.y < brick.y + brickHeight) {
                    ball.dy = -ball.dy;
                    brick.status = 0;
                    score += 10;
                    updateScore();
                    
                    // Check if all bricks are destroyed
                    if (score === brickRows * brickCols * 10) {
                        endGame('You Win! 🎉');
                    }
                }
            }
        }
    }
}

// Update game state
function update() {
    if (!gameRunning) return;

    // Move paddle
    if (rightPressed && paddle.x < canvas.width - paddle.width) {
        paddle.x += paddle.speed;
    } else if (leftPressed && paddle.x > 0) {
        paddle.x -= paddle.speed;
    }

    // Mouse control
    if (mouseX > 0) {
        paddle.x = mouseX - paddle.width / 2;
        if (paddle.x < 0) paddle.x = 0;
        if (paddle.x > canvas.width - paddle.width) paddle.x = canvas.width - paddle.width;
    }

    // Move ball
    ball.x += ball.dx;
    ball.y += ball.dy;

    // Ball collision with walls
    if (ball.x + ball.radius > canvas.width || ball.x - ball.radius < 0) {
        ball.dx = -ball.dx;
    }
    if (ball.y - ball.radius < 0) {
        ball.dy = -ball.dy;
    }

    // Ball collision with paddle
    if (ball.y + ball.radius > paddle.y &&
        ball.x > paddle.x && ball.x < paddle.x + paddle.width) {
        // Add some angle based on where ball hits paddle
        const hitPos = (ball.x - paddle.x) / paddle.width;
        ball.dx = ball.speed * (hitPos - 0.5) * 2;
        ball.dy = -Math.abs(ball.dy);
    }

    // Ball falls below paddle
    if (ball.y + ball.radius > canvas.height) {
        lives--;
        updateLives();
        if (lives === 0) {
            endGame('Game Over!');
        } else {
            resetBall();
        }
    }

    collisionDetection();
}

// Render game
function draw() {
    ctx.clearRect(0, 0, canvas.width, canvas.height);

    drawBricks();
    drawBall();
    drawPaddle();

    if (!gameStarted) {
        ctx.fillStyle = 'rgba(255, 255, 255, 0.8)';
        ctx.font = '24px Arial';
        ctx.textAlign = 'center';
        ctx.fillText('Press SPACE to Start!', canvas.width / 2, canvas.height / 2);
    }
}

// Game loop
function gameLoop() {
    update();
    draw();
    requestAnimationFrame(gameLoop);
}

// Game control functions
function startGame() {
    gameStarted = true;
    gameRunning = true;
}

function resetBall() {
    ball.x = canvas.width / 2;
    ball.y = canvas.height - 50;
    ball.dx = 5 * (Math.random() > 0.5 ? 1 : -1);
    ball.dy = -5;
    gameRunning = false;
    gameStarted = false;
}

function updateScore() {
    document.getElementById('score').textContent = score;
}

function updateLives() {
    document.getElementById('lives').textContent = lives;
}

function endGame(message) {
    gameRunning = false;
    document.getElementById('gameOverTitle').textContent = message;
    document.getElementById('finalScore').textContent = `Final Score: ${score}`;
    document.getElementById('gameOver').style.display = 'block';
}

function restartGame() {
    score = 0;
    lives = 3;
    gameRunning = false;
    gameStarted = false;

    updateScore();
    updateLives();

    initBricks();
    resetBall();

    document.getElementById('gameOver').style.display = 'none';
}

// Initialize game
initBricks();
gameLoop();
